### ISSUE TITLE
rspec output matcher no longer works with table_print

### ISSUE THREAD

**Issue by jonmchan** [arches/table_print#62]
In older versions of tableprint, one could write rspec stdout expectations such as:

```ruby
it { expect{ tp table }.to output('No data.').to_stdout }
```

But with the latest versions of table_print, we are using a separate stdout call (https://github.com/arches/table_print/blob/master/lib/table_print.rb#L73) than the kernel ```puts``` and the rspec output matcher is unable to intercept any of table_print's output making all tests against the library fail. 

Is there any way that we can work around this to allow other IOs to be specified as well as maintain compatible so that rspec output matchers can still work?

**Comment by jonmchan** [comment]
If it is may be of help, for reference, rspec stdout output matcher's implementation is:

https://github.com/rspec/rspec-expectations/blob/v3.1.2/lib/rspec/matchers/built_in/output.rb#L112

**Comment by blackbumer** [comment]
+1

**Comment by arches** [comment]
Last item here https://github.com/arches/table_print#config talks about using different IO objects, will that approach work for your scenario?

**Comment by arches** [comment]
Just thinking about this a little more.... I'm surprised it broke, [table_print's default IO is STDOUT](https://github.com/arches/table_print/blob/master/lib/table_print/config.rb#L12).

Also if there's a specific IO that would work better for you, I'd set it up in a before block or in your spec_helper: `tp.set :io, $stdout`

**Comment by jonmchan** [comment]
Hi @arches - thanks for looking into this. I think it is the implementation of rspec stdout matcher. If you look at the implementation, it looks like it is creating a new StringIO stream and assigning $stdout this new stream. I am going to assume put uses $stdout instead of STDOUT. When table_print prints to STDOUT, it isn't printing to rspec's $stdout stream.

        def self.capture(block)
          captured_stream = StringIO.new

          original_stream = $stdout
          $stdout = captured_stream

          block.call

          captured_stream.string
        ensure
          $stdout = original_stream
        end
      end

**Comment by arches** [comment]
hmmmmmmmmm that's going to be more difficult.  You can use the method I described above if you can set the tp IO object inside the rspec capture.  Do you know when exactly the capture is being set up?  Will you be able to set the tp IO object inside it?  Hopefully it's at least possible, and hopefully it doesn't mean you have to set the tp IO inside each test.  If there isn't a way to do it in a before block lmk and we can try to figure out something more clever.

**Comment by jonmchan** [comment]
I think the capture is being set up in the expect. I assume the block passed is exactly what you see in in the expect{}. The naive implementation would probably be something like this:

    it { expect{ tp.set :io, $stdout; tp table }.to output('No data.').to_stdout }

If we can set it earlier somehow or within a before block, that would be great. But as far as I understand, the capture is being generated right before calling the expect.

I'll test though if the above example works first though.

**Comment by jonmchan** [comment]
@arches i verified that adding a ```tp.set :io, $stdout``` within the expect block correctly sets the io to the correct captured_stream IO that the stdout rspec matcher is using. 

I also verified that trying to set tp to $stdout in a before block does not fix the issue. tp gets set to $stdout which is STDOUT at the time and then $stdout is changed and tp does not pickup this change.

**Comment by arches** [comment]
@jonmchan thanks for digging.  This is what I've learned about `STDOUT` vs `$stdout`:

https://stackoverflow.com/questions/6671716/what-is-the-difference-between-rubys-stdout-and-stdout

> $stdout is a global variable that represents the current standard output.  STDOUT is a constant representing standard output and is typically the default value of $stdout.

Sounds like I used the wrong one, so I expect a simple s/STDOUT/$stdout will fix this rspec problem with no adverse effects (*knock on wood*).  I'll try to get that done asap.
